netty/src/main/java/io/grpc/netty/GrpcHttp2HeadersDecoder.java

```diff
@@ -0,0 +1,498 @@
+      return str.charAt(0) == ':';
+    }
+
+    protected AsciiString validateName(AsciiString str) {
+      final int offset = str.arrayOffset();
+      final int length = str.length();
+      final byte[] data = str.array();
+      for (int i = offset; i < offset + length; i++) {
+        if (isUpperCase(data[i])) {
+          PlatformDependent.throwException(connectionError(PROTOCOL_ERROR,
+              "invalid header name '%s'", str));
+        }
+      }
+      return str;
+    }
+
+    private void expandHeadersAndValues() {
+      int oldLen = namesAndValues.length;
+      byte[][] newHeadersAndValues = new byte[oldLen + oldLen * 2][];
+      AsciiString[] newValues = new AsciiString[values.length + values.length * 2];
+      System.arraycopy(namesAndValues, 0, newHeadersAndValues, 0, oldLen);
+      System.arraycopy(values, 0, newValues, 0, values.length);
+      namesAndValues = newHeadersAndValues;
+      values = newValues;
+    }
+
+    @Override
+    public int size() {
+      return numHeaders();
+    }
+  }
+
+  /**
+   * A {@link GrpcHttp2InboundHeaders} implementation, optimized for HTTP/2 request headers. That
+   * is, HTTP/2 request pseudo headers are stored in dedicated fields and are NOT part of the
+   * array returned by {@link #namesAndValues()}.
+   *
+   * <p>This class only implements the methods used by {@link NettyServerHandler} and tests. All
+   * other methods throw an {@link UnsupportedOperationException}.
+   */
+  static final class GrpcHttp2RequestHeaders extends GrpcHttp2InboundHeaders {
+
+    private static final AsciiString PATH_HEADER = AsciiString.of(":path");
+    private static final AsciiString AUTHORITY_HEADER = AsciiString.of(":authority");
+    private static final AsciiString METHOD_HEADER = AsciiString.of(":method");
+    private static final AsciiString SCHEME_HEADER = AsciiString.of(":scheme");
+
+    private AsciiString path;
+    private AsciiString authority;
+    private AsciiString method;
+    private AsciiString scheme;
+    private AsciiString te;
+
+    GrpcHttp2RequestHeaders(int numHeadersGuess) {
+      super(numHeadersGuess);
+    }
+
+    @Override
+    public Http2Headers add(CharSequence csName, CharSequence csValue) {
+      final AsciiString name = validateName(requireAsciiString(csName));
+      final AsciiString value = requireAsciiString(csValue);
+      if (isPseudoHeader(name)) {
+        addPseudoHeader(name, value);
+        return this;
+      }
+      if (equals(TE_HEADER, name)) {
+        te = value;
+        return this;
+      }
+      return add(name, value);
+    }
+
+    @Override
+    public CharSequence get(CharSequence csName) {
+      final AsciiString name = requireAsciiString(csName);
+      if (isPseudoHeader(name)) {
+        // This code should never be reached.
+        throw new IllegalArgumentException("Use direct accessor methods for pseudo headers.");
+      }
+      if (equals(TE_HEADER, name)) {
```

Comment:
I can't imagine this getting executed often, but I would hope that pointer equality would be the preferred way of checking.